File: python/경주로건설.py
inf = 1000000000 #unsolved

def dp(board, dir, x, y, cost):
    
    board[x][y] = min(board[x][y], cost)
    if x == len(board)-1 and y == len(board)-1:
        return

    if dir == 'R':
        if y < len(board)-1 and board[x][y+1] != 1 and cost+100 <= board[x][y+1]:
            dp(board, dir, x, y+1, cost+100)
        if x < len(board)-1 and board[x+1][y] != 1 and cost+600 <= board[x+1][y]:
            dp(board, 'D', x+1, y, cost+600)
        if x > 0 and board[x-1][y] != 1 and cost+600 <= board[x-1][y]:
            dp(board, 'U', x-1, y, cost+600)
    if dir == 'L':
        if y > 0 and board[x][y-1] != 1 and cost+100 <= board[x][y-1]:
            dp(board, dir, x, y-1, cost+100)
        if x < len(board)-1 and board[x+1][y] != 1 and cost+600 <= board[x+1][y]:
            dp(board, 'D', x+1, y, cost+600)
        if x > 0 and board[x-1][y] != 1 and cost+600 <= board[x-1][y]:
            dp(board, 'U', x-1, y, cost+600)
    if dir == 'U':
        if x > 0 and board[x-1][y] != 1 and cost+100 <= board[x-1][y]:
            dp(board, dir, x-1, y, cost+100)
        if y < len(board)-1 and board[x][y+1] != 1 and cost+600 <= board[x][y+1]:
            dp(board, 'R', x, y+1, cost+600)
        if y > 0 and board[x][y-1] != 1 and cost+600 <= board[x][y-1]:
            dp(board, 'L', x, y-1, cost+600)
    if dir == 'D':
        if x < len(board)-1 and board[x+1][y] != 1 and cost+100 <= board[x+1][y]:
            dp(board, dir, x+1, y, cost+100)
        if y < len(board)-1 and board[x][y+1] != 1 and cost+600 <= board[x][y+1]:
            dp(board, 'R', x, y+1, cost+600)
        if y > 0 and board[x][y-1] != 1 and cost+600 <= board[x][y-1]:
            dp(board, 'L', x, y-1, cost+600)
    return board[len(board)-1][len(board)-1]

def solution(board):
    answer = 0
    for a in range(len(board)):
        for b in range(len(board)):
            if board[a][b] != 1:
                board[a][b] = inf
    
    answer = dp(board, 'R', 0, 0, 0)
    answer = dp(board, 'D', 0, 0, 0)
    print(board)

    return answer

#print(solution([[0,0,0],[0,0,0],[0,0,0]]))
#print(solution([[0,0,1,0],[0,0,0,0],[0,1,0,1],[1,0,0,0]]))
#print(solution([[0, 0, 1, 0, 1, 1, 0, 0, 0, 0], [0, 0, 0, 0, 1, 0, 1, 1, 0, 1], [1, 0, 0, 0, 0, 1, 1, 0, 1, 0], [0, 0, 0, 0, 0, 0, 1, 0, 0, 0], [0, 0, 0, 0, 1, 0, 1, 0, 1, 1], [0, 0, 1, 0, 1, 1, 0, 1, 0, 1], [0, 1, 0, 0, 1, 0, 0, 0, 1, 0], [1, 0, 0, 1, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 1, 0, 1, 0, 0], [1, 0, 0, 0, 0, 0, 0, 0, 1, 0]]))
print(solution([[0, 1, 1, 1, 1, 1, 1, 1, 1],
[0, 0, 0, 0, 0, 1, 1, 1, 1],
[1, 1, 1, 1, 0, 1, 1, 1, 1],
[0, 0, 0, 0, 0, 1, 1, 1, 1],
[0, 1, 0, 1, 1, 1, 1, 1, 1],
[0, 1, 0, 0, 0, 1, 1, 1, 1],
[0, 1, 1, 1, 0, 1, 1, 1, 1],
[0, 0, 0, 0, 0, 1, 1, 1, 1],
[1, 1, 1, 1, 0, 0, 0, 0, 0]]))

File: JavaScript/경주로건설.js
const getCost = (board, costBoard, i, j, cost, vertical, horizontal) => {
    board[i][j] = 1;
    costBoard[i][j] = cost;
    const destination = board.length - 1;
    if (i === destination && j === destination) {
        return cost;
    }
    let answer = 10e9;
    if (j > 0 && board[i][j - 1] !== 1) {
        const additionalCost = horizontal ? 600 : 100;
        if (costBoard[i][j - 1] + 250 >= cost + additionalCost) {
            answer = Math.min(answer, getCost(JSON.parse(JSON.stringify(board)), costBoard, i, j - 1, cost + additionalCost, true, false));
        }
    }
    if (j < destination && board[i][j + 1] !== 1) {
        const additionalCost = horizontal ? 600 : 100;
        if (costBoard[i][j + 1] + 250 >= cost + additionalCost) {
            answer = Math.min(answer, getCost(JSON.parse(JSON.stringify(board)), costBoard, i, j + 1, cost + additionalCost, true, false));
        }
    }
    if (i > 0 && board[i - 1][j] !== 1) {
        const additionalCost = vertical ? 600 : 100;
        if (costBoard[i - 1][j] + 250 >= cost + additionalCost) {
            answer = Math.min(answer, getCost(JSON.parse(JSON.stringify(board)), costBoard, i - 1, j, cost + additionalCost, false, true));
        }
    }
    if (i < destination && board[i + 1][j] !== 1) {
        const additionalCost = vertical ? 600 : 100;
        if (costBoard[i + 1][j] + 250 >= cost + additionalCost) {
            answer = Math.min(answer, getCost(JSON.parse(JSON.stringify(board)), costBoard, i + 1, j, cost + additionalCost, false, true));
        }
    }
    return answer;
}

function solution(board) {
    let newBoard = JSON.parse(JSON.stringify(board));
    let costBoard = new Array(board.length);
    for (let i = 0; i < board.length; i++) {
        const costs = new Array(board.length).fill(10e9);
        costBoard[i] = costs;
    }
    return getCost(newBoard, costBoard, 0, 0, 0, false, false);
}

//console.log(solution([[0, 0, 0], [0, 0, 0], [0, 0, 0]]));
console.log(solution([
    [0, 0, 0, 0, 0, 0, 0, 1], 
    [0, 0, 0, 0, 0, 0, 0, 0], 
    [0, 0, 0, 0, 0, 1, 0, 0], 
    [0, 0, 0, 0, 1, 0, 0, 0], 
    [0, 0, 0, 1, 0, 0, 0, 1], 
    [0, 0, 1, 0, 0, 0, 1, 0], 
    [0, 1, 0, 0, 0, 1, 0, 0], 
    [1, 0, 0, 0, 0, 0, 0, 0]
]));